client/state/data-layer/http-data/index.js

```diff
@@ -0,0 +1,66 @@
+/** @format */
+/**
+ * Internal dependencies
+ */
+import { HTTP_DATA_REQUEST, HTTP_DATA_TICK } from 'state/action-types';
+import { dispatchRequestEx } from 'state/data-layer/wpcom-http/utils';
+import { update } from './common';
+
+const fetch = action => {
+	update( action.id, 'pending' );
+
+	return [
+		{
+			...action.fetch,
+			onSuccess: action,
+			onFailure: action,
+			onProgress: action,
+		},
+		{ type: HTTP_DATA_TICK },
+	];
+};
+
+const onFailure = ( action, error ) => {
+	update( action.id, 'failure', error );
+
+	return { type: HTTP_DATA_TICK };
+};
+
+const parseResponse = ( data, fromApi ) => {
+	if ( 'function' !== typeof fromApi ) {
+		return data;
+	}
+
+	// [ error, data ]
+	try {
+		return [ undefined, fromApi( data ) ];
+	} catch ( error ) {
+		return [ error, undefined ];
+	}
+};
+
+const onSuccess = ( action, apiData ) => {
+	const [ error, data ] = parseResponse( apiData, action.fromApi );
+
+	if ( undefined !== error ) {
+		return onFailure( action, error );
+	}
+
+	if ( 'multi-resource' === action.fromApiType ) {
```

Comment:
do you already have a use-case that would benefit from `multi-resource`?

Reply:
getting any collection of things like posts or comments or activity log items.

I have decided that we should update the original request identifier too

component wants to know "have I fetched posts for this site yet?" and
the request id shows pending then done. however, it brings in twenty
posts. those posts may or may not be in memory when the post component
displays it. so it says, "have I fetched this post yet?" and now because
we requested it then the post is there.

also this means that the post shows up in memory whether we fetched it
from "get all posts for site" or "fetch this specific post" - it's part-and-parcel
of how the data types and ids work.